// sanity/lib/_legacy/generate-help-url.esm.esm.js
import { jsx, jsxs } from "react/jsx-runtime";
import { Button as Button$1, Dialog as Dialog$1, Flex, Box, MenuButton as MenuButton$1, MenuGroup as MenuGroup$1, Text, Stack, Hotkeys, Badge, MenuItem as MenuItem$1, Popover as Popover$1, Tab as Tab$1, Tooltip as Tooltip$1, TooltipDelayGroupProvider as TooltipDelayGroupProvider$1 } from "@sanity/ui";
import { forwardRef, useCallback, useMemo, isValidElement, createElement } from "react";
import { isValidElementType } from "react-is";
import { styled } from "styled-components";
import { useTranslation } from "react-i18next";
function ConditionalWrapper({
  children,
  condition,
  wrapper
}) {
  return condition ? wrapper(children) : children;
}
const LARGE_BUTTON_PROPS = {
  space: 3,
  padding: 3
}, DEFAULT_BUTTON_PROPS = {
  space: 2,
  padding: 2
}, TooltipButtonWrapper = styled.span`
  display: inline-flex;
`, Button = forwardRef(function({
  size = "default",
  mode = "default",
  paddingY,
  tone = "default",
  tooltipProps,
  ...rest
}, ref) {
  const renderWrapper = useCallback(
    (children) => /* @__PURE__ */ jsx(Tooltip, { content: tooltipProps == null ? void 0 : tooltipProps.content, portal: !0, ...tooltipProps, children: /* @__PURE__ */ jsx(TooltipButtonWrapper, { children }) }),
    [tooltipProps]
  );
  return /* @__PURE__ */ jsx(ConditionalWrapper, { condition: !!tooltipProps, wrapper: renderWrapper, children: /* @__PURE__ */ jsx(Button$1, { ...rest, ...size === "default" ? DEFAULT_BUTTON_PROPS : LARGE_BUTTON_PROPS, paddingY, ref, mode, tone }) });
}), Dialog = forwardRef(function({
  bodyHeight,
  children,
  footer,
  padding = !0,
  ...props
}, ref) {
  const { t } = useTranslation();
  return /* @__PURE__ */ jsx(
    Dialog$1,
    {
      ...props,
      animate: !0,
      ref,
      footer: ((footer == null ? void 0 : footer.confirmButton) || (footer == null ? void 0 : footer.cancelButton)) && /* @__PURE__ */ jsxs(Flex, { width: "full", gap: 3, justify: "flex-end", padding: 3, children: [
        props.onClose && /* @__PURE__ */ jsx(
          Button$1,
          {
            mode: "bleed",
            padding: 2,
            text: t("common.dialog.cancel-button.text"),
            tone: "default",
            onClick: props.onClose,
            ...footer.cancelButton
          }
        ),
        footer.confirmButton && /* @__PURE__ */ jsx(
          Button$1,
          {
            mode: "default",
            padding: 2,
            text: t("common.dialog.confirm-button.text"),
            tone: "critical",
            ...footer.confirmButton
          }
        )
      ] }),
      children: /* @__PURE__ */ jsx(Box, { height: bodyHeight, padding: padding ? 4 : 0, children })
    }
  );
}), MenuButton = forwardRef(function(props, ref) {
  return /* @__PURE__ */ jsx(
    MenuButton$1,
    {
      ...props,
      ref,
      popover: {
        ...props.popover,
        animate: !0
      }
    }
  );
}), MenuGroup = (props) => /* @__PURE__ */ jsx(MenuGroup$1, { ...props, fontSize: 1, padding: 3 }), FONT_SIZE = 1, SUBTITLE_FONT_SIZE = 0, SubtitleText = styled(Text)`
  margin-top: 2px;
`, PreviewWrapper = styled(Box)`
  height: 25px;
  width: 25px;
  overflow: hidden;
`, MenuItem = forwardRef(function({
  badgeText,
  children: childrenProp,
  disabled,
  hotkeys,
  icon,
  iconRight,
  preview = null,
  renderMenuItem,
  text,
  tooltipProps,
  __unstable_subtitle,
  __unstable_space,
  ...rest
}, ref) {
  const menuItemContent = useMemo(() => /* @__PURE__ */ jsxs(Flex, { align: "center", gap: 2, children: [
    preview && /* @__PURE__ */ jsx(
      PreviewWrapper,
      {
        style: { opacity: disabled ? 0.25 : void 0 },
        paddingRight: __unstable_space ? 1 : 0,
        children: /* @__PURE__ */ jsx(Flex, { align: "center", height: "fill", justify: "center", children: preview })
      }
    ),
    icon && /* @__PURE__ */ jsx(Box, { paddingRight: 1, children: /* @__PURE__ */ jsxs(Text, { size: FONT_SIZE, children: [
      isValidElement(icon) && icon,
      isValidElementType(icon) && createElement(icon)
    ] }) }),
    text && /* @__PURE__ */ jsxs(
      Stack,
      {
        flex: 1,
        space: __unstable_subtitle ? 1 : 2,
        paddingLeft: __unstable_subtitle ? 1 : 0,
        children: [
          /* @__PURE__ */ jsx(Text, { size: FONT_SIZE, textOverflow: "ellipsis", weight: "medium", children: text }),
          __unstable_subtitle && /* @__PURE__ */ jsx(SubtitleText, { size: SUBTITLE_FONT_SIZE, textOverflow: "ellipsis", weight: "medium", muted: !0, children: __unstable_subtitle })
        ]
      }
    ),
    (badgeText || hotkeys || iconRight) && /* @__PURE__ */ jsxs(Flex, { align: "center", gap: 3, marginLeft: 3, children: [
      hotkeys && /* @__PURE__ */ jsx(Hotkeys, { keys: hotkeys, style: { marginTop: -4, marginBottom: -4 } }),
      badgeText && /* @__PURE__ */ jsx(Badge, { fontSize: 0, style: { marginTop: -4, marginBottom: -4 }, children: badgeText }),
      iconRight && /* @__PURE__ */ jsxs(Text, { size: FONT_SIZE, children: [
        isValidElement(iconRight) && iconRight,
        isValidElementType(iconRight) && createElement(iconRight)
      ] })
    ] })
  ] }), [
    preview,
    disabled,
    __unstable_space,
    icon,
    text,
    __unstable_subtitle,
    badgeText,
    hotkeys,
    iconRight
  ]), renderWrapper = useCallback(
    (children) => /* @__PURE__ */ jsx(Tooltip, { content: tooltipProps == null ? void 0 : tooltipProps.content, portal: !0, ...tooltipProps, children: /* @__PURE__ */ jsx("div", { children }) }),
    [tooltipProps]
  );
  return /* @__PURE__ */ jsx(ConditionalWrapper, { condition: !!tooltipProps, wrapper: renderWrapper, children: /* @__PURE__ */ jsx(
    MenuItem$1,
    {
      disabled,
      paddingLeft: preview ? 1 : 3,
      paddingRight: 3,
      paddingY: preview ? 1 : 3,
      ref,
      ...rest,
      children: typeof childrenProp > "u" && typeof renderMenuItem == "function" ? renderMenuItem(menuItemContent) : menuItemContent
    }
  ) });
}), Popover = forwardRef(function(props, ref) {
  return /* @__PURE__ */ jsx(Popover$1, { ...props, animate: !0, ref });
}), Tab = forwardRef(function({ tone = "default", ...props }, ref) {
  return /* @__PURE__ */ jsx(Tab$1, { ...props, muted: !0, padding: 2, ref, tone });
}), TOOLTIP_DELAY_PROPS = {
  open: 400
}, TOOLTIP_SHARED_PROPS = {
  animate: !0,
  arrow: !1,
  boundaryElement: null,
  delay: TOOLTIP_DELAY_PROPS,
  fallbackPlacements: ["bottom-start", "bottom-end", "top-start", "top-end"],
  placement: "bottom",
  portal: !0
}, Tooltip = forwardRef(function(props, ref) {
  const { content, hotkeys, ...rest } = props;
  return typeof content == "string" ? /* @__PURE__ */ jsx(
    Tooltip$1,
    {
      ...TOOLTIP_SHARED_PROPS,
      content: /* @__PURE__ */ jsxs(Flex, { align: "center", children: [
        content && /* @__PURE__ */ jsx(Box, { flex: 1, padding: 1, children: /* @__PURE__ */ jsx(Text, { size: 1, children: content }) }),
        hotkeys && /* @__PURE__ */ jsx(Box, { flex: "none", children: /* @__PURE__ */ jsx(Hotkeys, { keys: hotkeys }) })
      ] }),
      padding: 1,
      ref,
      ...rest
    }
  ) : /* @__PURE__ */ jsx(Tooltip$1, { ...TOOLTIP_SHARED_PROPS, content, ref, ...rest });
}), TooltipDelayGroupProvider = (props) => /* @__PURE__ */ jsx(TooltipDelayGroupProvider$1, { delay: TOOLTIP_DELAY_PROPS, children: props.children }), BASE_URL = "https://docs.sanity.io/help/";
function generateHelpUrl(slug) {
  return BASE_URL + slug;
}
export {
  Button,
  ConditionalWrapper,
  Dialog,
  MenuButton,
  MenuGroup,
  MenuItem,
  Popover,
  Tab,
  Tooltip,
  TooltipDelayGroupProvider,
  generateHelpUrl
};
